src/tequila/apps/qcc/qcc.py

```diff
@@ -0,0 +1,394 @@
+# Qubit Coupled Cluster
+# as described in Ryabinkin, Yen, Genin, Izmaylov: https://doi.org/10.1021/acs.jctc.8b00932
+# and Iterative Qubit Coupled Cluster
+# as described in Ryabinkin, Lang, Genin, Izmaylov: https://doi.org/10.1021/acs.jctc.9b01084
+
+import itertools
+import numpy as np
+from openfermion import QubitOperator
+
+import tequila as tq
+from tequila.circuit import QCircuit
+from tequila.hamiltonian.qubit_hamiltonian import BinaryPauli
+from tequila.grouping.binary_utils import binary_phase
+from tequila.grouping.binary_rep import BinaryHamiltonian, BinaryPauliString
+
+
+def are_equal_bitwise_check(binary_vec1, binary_vec2):
+    """
+    Checks if two binary vectors identical with O(log(n)) average case
+    run-time.
+
+    Arguments:
+        binary_vec1 (array_like): the first binary vector
+        binary_vec2 (array_like): the second binary vector
+
+    Returns:
+        bool: True if they are the same vector, False otherwise.
+    """
+    n = len(binary_vec1)
+    for i in range(n):
+        if binary_vec1[i] != binary_vec2[i]:
+            return False
+    return True
+
+
+def z_eigenvalue(binary_occupations, binary_z_placements):
+    """
+    Computes the eigenvalue of a tensor product of Pauli Z operators with
+    respect to a computational basis state.
+
+    Arguments:
+        binary_occupations (array_like): binary vector representing a computational basis
+            state.
+        binary_z_placements (array_like): binary vector with 1 in indices within the
+            support of the tensor product of Pauli Z operators, with 0 elsewhere.
+
+    Returns:
+        int: the eigenvalue value 1 or -1.
+    """
+    return 1 - 2 * (len(np.where(binary_occupations + binary_z_placements == 2)[0]) % 2)
+
+
+def even_x_y_overlap_phase(z_vector, x_vector):
+    """
+    Convenience function used in performing the Ising factorization of the Hamiltonian.
+    This function computes the phase resulting from multiplying a tensor product of X
+    Pauli operations with a tensor product of Z Pauli operations, X*Z, where it is a priori
+    known that the two tensor products have even overlapping support.
+
+    Arguments:
+        z_vector (array_like): binary vector with 1 in indices the Z operator has
+            support on, with 0 elsewhere.
+
+        x_vector (array_like): binary vector with 1 in indices the X operator has
+            support on, with 0 elsewhere.
+
+    Returns:
+        int: 1 or -1.
+    """
+    return 1 - 2 * ((len(np.where(x_vector + z_vector == 2)[0]) // 2) % 2)
+
+
+class IterativeQCC:
+    """
+    Class for performing various computations involved in the iterative QCC method,
+    along with full QCC and iterative QCC ground state energy estimations.
+    Directly following instantiation, the class method `do_iteration` can be called
+    to perform a complete iQCC iteration, while `do_qcc` can be used to perform a one-shot
+    QCC calculation.
+    """

```

Comment:
Would be good to have a call operator, like
def __call__(iterative: bool = False, *args, **kwargs):
    if iterative:
        ...
    else:
        return self.do_iteration()

that performs the default operation (would suggest standard QCC; just for convenience)
Then you can just do:

solver = IterativeQCC(....)
result = solver()